ignite/engine/state.py

```diff
@@ -64,18 +79,34 @@ def __init__(self, **kwargs: Any) -> None:
         for k, v in kwargs.items():
             setattr(self, k, v)
 
-        self._update_attrs()
+        # If it's a standalone state with no engine provided, then we must update
+        # all the attributes and set all initial values to zeros,
+        # epoch and iteration included. We do this to keep BC compatibility,
+        # as we now don't keep iteration and epoch here in state,
+        # instead of that, we use engine._allowed_events_counts to keep track of them.
+        if self.engine is None:
+            self._update_attrs()
+
+    # @property
+    # def event_to_attr(self) -> event_to_attr:
+    #     warnings.warn("'event_to_attr' is deprecated, please use 'attr_to_evets' instead.")
+    #     return State.event_to_attr
 
     def _update_attrs(self) -> None:
-        for value in self.event_to_attr.values():
-            if not hasattr(self, value):
-                setattr(self, value, 0)
+        for key in self.attr_to_events.keys():
+            if not hasattr(self, key):
+                setattr(self, key, 0)
 
     def get_event_attrib_value(self, event_name: Union[str, Events, CallableEventWithFilter]) -> int:
         """Get the value of Event attribute with given `event_name`."""
-        if event_name not in State.event_to_attr:
+        event_to_attr = {}
```

Comment:
Shouldn't we deprecate this method as well ?

Reply:
I think we should leave this till we find a replacement, check this please https://github.com/pytorch/ignite/pull/2128#discussion_r684754648

Reply:
I think, this method can go to EventsDrivenState and fetch the value from Engine. If no engine, raise an error.

Reply:
Thanks for this comment, this comment pointed me to a dangerous bug, I will try to explain it here
If something like this happened 
```py
engine = Engine(...)
engine.state = State(...)
```
normally we set the new state attributes only, and set engine to self. But the bug was that:
- `state._attr_to_events` is now None, which is wrong.
- `allowed_events_count` is not set to the new values of the attributes.
I discovered that when I was trying to get move this method to `EventsDrivenState` and get the values vai `allowed_events_counts` to drop `event_to_attr` totally. I'm working on fixing this now.

Reply:
My thoughts on this:
- for fixing the `state._attr_to_evets` issue, I'm thinking about passing the default `attr_to_events` driven in `super()` while initializing state if `state._attr_to_events` is None.
- for fixing `allowed_events_counts` I'm trying different solutions now.

Reply:
Almost solved it, another point about changing the code `get_event_attrib_value`, We have to keep the code the same way for now, in all tests we mock the engine, so `state.engine` is none, so we can't use `allowed_events_..` to get anything.